core/src/trezor/workflow.py

```diff
@@ -130,3 +135,70 @@ def _finalize_default(task: loop.Task, value: Any) -> None:
 # If required, a function `shutdown_default` should be written, that clears the
 # default constructor and shuts down the running default task.
 # We currently do not need such function, so I'm just noting how it should work.
+
+
+class IdleTimer:
+    """Run callbacks after a period of inactivity.
+
+    A global instance `workflow.idle_timer` is available to create events that fire
+    after a specified time of no user or host activity. This instance is kept awake
+    by UI taps, swipes, and USB message handling.
+    """
+
+    def __init__(self) -> None:
+        self.timeouts = {}  # type: Dict[IdleCallback, int]
+        self.tasks = {}  # type: Dict[IdleCallback, loop.Task]
+
+    async def _timeout_task(self, callback: IdleCallback) -> None:
+        # This function is async, so the result of self._timeout_task() is an awaitable,
+        # suitable for scheduling.
+
+        # After the scheduled task completes, self.tasks will contain a stale task
+        # object. A new one must be created here so that subsequent calls to touch() can
+        # schedule it again.
+        self.tasks[callback] = self._timeout_task(callback)
+        callback()
+
+    def touch(self) -> None:
+        """Wake up the idle timer.
+
+        Events that represent some form of activity (USB messages, touches, etc.) should
+        call `touch()` to notify the timer of the activity. All pending callback timers
+        will reset.
+        """
+        for callback, task in self.tasks.items():
+            timeout_us = self.timeouts[callback]
+            deadline = utime.ticks_add(utime.ticks_us(), timeout_us)
+            loop.schedule(task, None, deadline, reschedule=True)
+
+    def set(self, timeout_ms: int, callback: IdleCallback) -> None:
+        """Add or update an idle callback.
+
+        Every time `timeout_ms` milliseconds elapse after the last registered activity,
+        `callback` will be invoked.
+        I.e., in every period of inactivity, each `callback` will only run once. To run
+        again, an activity must be registered and then no activity for the specified
+        period.
+
+        If `callback` was previously registered, it is updated with a new timeout value.
+
+        `idle_timer.set()` also counts as an activity, so all running idle timers are
+        reset.
+        """
+        if callback in self.tasks:
+            loop.close(self.tasks[callback])
+
+        self.timeouts[callback] = timeout_ms * 1000
+        self.tasks[callback] = self._timeout_task(callback)
+        self.touch()
```

Comment:
TBH I don't understand what's going on here, but why should set() call self.touch()? I thought touch() resets the idle timer. It just seems surprising that adding a callback would do that.

Reply:
It resolves an ambiguity wrt idleness and relative ordering of the idle tasks.

Namely: if `self.touch()` didn't happen, should this schedule the task to 10 minutes from _now_, or from _last touch_? If from _now_, other tasks scheduled to 10 minutes might happen at different time (but only for the first time, next time they will be the same). If from _last touch_, and more than 10 minutes elapsed, should the task run right now, or only next time an activity is registered and 10 minutes elapse?

I mean, all of this is pointless because adding a task only happens _during_ an activity. But counting `set()` as an activity seems very clean.
We can easily change this if there are new requirements.

Reply:
i'll add a comment to the code about this

Reply:
added in 20d7fe0